Add functionality to retrieve input source from MiniDSP 2x4HD

dennisfrett opened the pull request:
Additional functionality for the MiniDSP driver.

Next to volume and muted status, it can now also retrieve the current input source (Analog, Toslink or USB).

Lauszus commented on examples/MiniDSP/MiniDSP.ino:
```diff
@@ -25,6 +25,20 @@ void OnMutedChange(bool isMuted) {
 }
 
+void OnInputSourceChange(MiniDSP::InputSource inputSource) {
+  String inputSourceStr;
```
Maybe initialize this to "Unknown"?

Lauszus commented:
Looking good. I've left one comment. Can you address this and then I'll merge it :)

dennisfrett commented:
> I added also the functionality to retrieve current config from MiniDSP 2x4HD.

I can check if I can add these later this week. @Lauszus I don't know if you'd rather merge this change in already, or if you'd like to merge everything together later?

MarkLido commented:
Better later. After the function for "send" is added!?

Lauszus commented:
Okay great. Just let me know when it's ready to get merged :)

dennisfrett commented:
I updated the code to also retrieve the current config + I fixed the callback for input source not being called (Might be handy for you, @MarkLido ).

Updating volume has not been done yet.

However, @Lauszus , I'm having an issue, in `OnInitSuccessful` I do:
```lang=cpp
        RequestStatus();
        RequestInputSource();
        RequestConfig();
```

Which seems to be sending commands, but not listening for a response,  so most of the time the response to the first command (request status), is just lost.

What would you suggest here?

I guess I'm using the library wrong, should I use `ctrlReq` instead? Or should I somehow busy wait for a response myself?

MarkLido commented:
Thanks Dennis. The callback for input source works now.

dennisfrett commented:
@MarkLido I've made a quick implementation for setting volume and muted status.

But I still think I might not be handling synchronous commands correctly, so any input from @Lauszus here would be awesome.

Lauszus commented on MiniDSP.cpp:
```diff
@@ -109,3 +163,33 @@ void MiniDSP::RequestStatus() const {
+        uint8_t SetMutedommand[] = {0x17, muted ? (uint8_t)0x01 : (uint8_t)0x00};
+
+        SendCommand(SetMutedommand, sizeof(SetMutedommand));
+}
```
Missing new line.

Lauszus commented on MiniDSP.cpp:
```diff
@@ -109,3 +163,33 @@ void MiniDSP::RequestStatus() const {
+                return;
+        }
+
+        uint8_t SetVolumeCommand[] = {0x42, (uint8_t)(-2*volumeDB)};
```
```suggestion
       uint8_t SetVolumeCommand[] = {0x42, (uint8_t)round(-2.0f * volumeDB)};
```

Lauszus commented on MiniDSP.cpp:
```diff
@@ -109,3 +163,33 @@ void MiniDSP::RequestStatus() const {
+void MiniDSP::setVolumeDB(float volumeDB) const {
+        // Only accept values between 0dB and -127.5dB.
+        // Don't do error handling.
+        if(volume > 0 || volume < -127.5){
```
```suggestion
        if(volume < 0 || volume > -127.5) {
```

Lauszus commented on MiniDSP.h:
```diff
@@ -159,6 +219,18 @@ class MiniDSP : public HIDUniversal {
+        /**
+         * Send the "Request input source" command to the MiniDSP.
+         */
+        void
```
```suggestion
        void RequestInputSource() const;
```

Lauszus commented on MiniDSP.h:
```diff
@@ -159,6 +219,18 @@ class MiniDSP : public HIDUniversal {
+        /**
+         * Send the "Request config" command to the MiniDSP.
+         */
+        void
```
```suggestion
        void RequestConfig() const;
```

Lauszus commented:
> I updated the code to also retrieve the current config + I fixed the callback for input source not being called (Might be handy for you, @MarkLido ).
> 
> Updating volume has not been done yet.
> 
> However, @Lauszus , I'm having an issue, in `OnInitSuccessful` I do:
> 
> ```
>         RequestStatus();
>         RequestInputSource();
>         RequestConfig();
> ```
> 
> Which send commands, but don't listening for a response, so most of the time the response to the first command (request status), is just lost.
> 
> What would you suggest here?
> 
> I guess I'm using the library wrong, should I use `ctrlReq` instead? Or should I somehow busy wait for a response myself?

An workaround would be to just add `delay(1);` in between the calls.

dennisfrett commented:
@Lauszus Writing this driver without a way too easily debug USB pakets seems quite difficult, and my current implementation has some bugs (`delay`ing inbetween sending doesn't resolve the issue, setting volume only works the first time and not the second time, ...). To make my life easier I've purchased an Atmega32u4 in the hope I can set it up as a USB sniffer (https://github.com/matlo/serialusb). 

Once I figure out what's going wrong and fix it, I'll update the PR.
